tsuru/client/build.go

```diff
@@ -0,0 +1,162 @@
+package client
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/tsuru/gnuflag"
+	"github.com/tsuru/tsuru/cmd"
+	tsuruIo "github.com/tsuru/tsuru/io"
+	"github.com/tsuru/tsuru/safe"
+)
+
+type AppBuild struct {
+	cmd.GuessingCommand
+	tag string
+	fs  *gnuflag.FlagSet
+}
+
+func (c *AppBuild) Flags() *gnuflag.FlagSet {
+	if c.fs == nil {
+		c.fs = c.GuessingCommand.Flags()
+		tag := "The image tag"
+		c.fs.StringVar(&c.tag, "tag", "", tag)
+		c.fs.StringVar(&c.tag, "t", "", tag)
+	}
+	return c.fs
+}
+
+func (c *AppBuild) Info() *cmd.Info {
+	desc := `Builds a tsuru app image. Some examples of calls are:
+
+::
+
+		$ tsuru app-build -a myapp -t mytag .
+		$ tsuru app-build -a myapp -t latest myfile.jar Procfile
+`
+	return &cmd.Info{
+		Name:    "app-build",
+		Usage:   "app-build [-a/--app <appname>] [-t/--tag <image_tag>] <file-or-dir-1> [file-or-dir-2] ... [file-or-dir-n]",
+		Desc:    desc,
+		MinArgs: 0,
+	}
+}
+
+func (c *AppBuild) Run(context *cmd.Context, client *cmd.Client) error {
+	context.RawOutput()
+	if c.tag == "" {
```

Comment:
Shouldn't you validate the tag (like checking for invalid characters)? Or do you leave this validation to the server?

Reply:
I will do this on server side